scripts/apogee/build_parent_sample.py

```diff
@@ -0,0 +1,283 @@
+        base_path,
+        f"spectro/aspcap/dr17/{aspcap_code}/{telescope}",
+        str(f"{field}"),
+    )
+    if not os.path.exists(fullfoldername):
+        os.makedirs(fullfoldername)
+
+    fullfilename = os.path.join(fullfoldername, filename)
+
+    if os.path.exists(fullfilename):
+        return fullfilename
+    else:
+        try:
+            urllib.request.urlretrieve(urlstr, fullfilename)
+            return 0
+        except urllib.error.HTTPError as emsg:
+            return 1  # error code
+
+
+def visit_spectra(
+    base_path,
+    field,
+    telescope,
+    filename,
+):
+    str1 = f"https://data.sdss.org/sas/dr17/apogee/spectro/redux/dr17/stars/{telescope}/{field}/"
+    urlstr = str1 + filename
+
+    fullfoldername = os.path.join(
+        base_path,
+        f"spectro/redux/dr17/stars/{telescope}/",
+        str(f"{field}"),
+    )
+
+    if not os.path.exists(fullfoldername):
+        os.makedirs(fullfoldername)
+
+    fullfilename = os.path.join(fullfoldername, filename)
+    if os.path.exists(fullfilename):
+        return fullfilename
+    else:
+        try:
+            urllib.request.urlretrieve(urlstr, fullfilename)
+            return 0
+        except urllib.error.HTTPError as emsg:
+            return 1  # error code
+
+
+def selection_fn(base_path, catalog):
+    # Only use the spectrum from APO 2.5m and LCO 2.5m
+    mask = (catalog["TELESCOPE"] == "apo25m") | (catalog["TELESCOPE"] == "lco25m")
+    # known no file entries
+    mask &= ~catalog["FILE"].mask
+
+    # went thru all to check if those exist
+    for idx, (apogee_id, field, telescope, filename) in enumerate(zip(catalog["APOGEE_ID"], catalog["FIELD"], catalog["TELESCOPE"], catalog["FILE"])):
+        if not mask[idx] or filename is None:  # no need to do anything if already bad
+            mask[idx] = False
+            continue
+        result = visit_spectra(base_path, field, telescope, filename)
+        if result == 1:
+            mask[idx] = False
+            continue
+        result = combined_spectra(base_path, field, apogee_id, telescope)
+        if result == 1:
+            mask[idx] = False
+    return mask
+
+
+def processing_fn(raw_filename, continuum_filename):
+    """Parallel processing function reading all requested spectra from one plate."""
+
+    # Load the visit spectra file
+    if os.path.exists(raw_filename) and raw_filename != 1:
+        hdus = fits.open(raw_filename)
+        raw_flux = hdus[1].data[0]
+        raw_ivar = 1 / hdus[2].data[0] ** 2
+        mask_spec = hdus[2].data[0]
+    else:
+        raw_flux = np.zeros(8575)
+        raw_ivar = np.zeros(8575)
+        mask_spec = np.zeros(8575)
+
+    # Load the combined spectra file
+    if os.path.exists(continuum_filename) and continuum_filename != 1:
+        hdus = fits.open(continuum_filename)
+        continuum_flux = hdus[1].data
+        continuum_ivar = 1 / hdus[2].data ** 2
+    else:
+        continuum_flux = np.zeros(8575)
```

Comment:
does this mean that we can have empty entries? Can we find those at the catalog level and remove them with some sort of selection?

Reply:
Actually the selection function have filtered out those empty entries so I have removed these codes now